storage/aptosdb/src/db_debugger/truncate/mod.rs

```diff
@@ -0,0 +1,567 @@
+            iter.seek(&NodeKey::new_empty_path(current_version))?;
+            for item in iter {
+                let (key, _) = item?;
+                batch.delete::<JellyfishMerkleNodeSchema>(&key)?;
+            }
+
+            Self::delete_stale_node_index_at_version::<StaleNodeIndexSchema>(
+                state_merkle_db,
+                current_version,
+                &batch,
+            )?;
+            Self::delete_stale_node_index_at_version::<StaleNodeIndexCrossEpochSchema>(
+                state_merkle_db,
+                current_version,
+                &batch,
+            )?;
+
+            state_merkle_db.write_schemas(batch)?;
+        }
+
+        Ok(())
+    }
+
+    fn delete_stale_node_index_at_version<S>(
+        state_merkle_db: &DB,
+        version: Version,
+        batch: &SchemaBatch,
+    ) -> Result<()>
+    where
+        S: Schema<Key = StaleNodeIndex>,
+        Version: SeekKeyCodec<S>,
+    {
+        let mut iter = state_merkle_db.iter::<S>(ReadOptions::default())?;
+        iter.seek(&version)?;
+        for item in iter {
+            let (index, _) = item?;
+            assert_eq!(index.stale_since_version, version);
+            batch.delete::<S>(&index)?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{
+        test_helper::{arb_blocks_to_commit_with_block_nums, update_in_memory_state},
+        AptosDB,
+    };
+    use aptos_storage_interface::{DbReader, DbWriter};
+    use aptos_temppath::TempPath;
+    use proptest::prelude::*;
+
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(10))]
+
+        #[test]
+        fn test_truncation(input in arb_blocks_to_commit_with_block_nums(5, 15)) {
+            let tmp_dir = TempPath::new();
+            let db = AptosDB::new_for_test(&tmp_dir);
+            let mut in_memory_state = db.state_store.buffered_state().lock().current_state().clone();
+            let _ancestor = in_memory_state.base.clone();
+            let mut version = 0;
+            for (txns_to_commit, ledger_info_with_sigs) in input.iter() {
+                update_in_memory_state(&mut in_memory_state, txns_to_commit.as_slice());
+                db.save_transactions(txns_to_commit, version, version.checked_sub(1), Some(ledger_info_with_sigs), true, in_memory_state.clone())
+                    .unwrap();
+                version += txns_to_commit.len() as u64;
+            }
+
+            let db_version = db.get_latest_transaction_info_option().unwrap().unwrap().0;
+            prop_assert_eq!(db_version, version - 1);
+
+            drop(db);
+
+            let target_version = db_version - 5;
+
+            let cmd = Cmd {
+                db_dir: tmp_dir.path().to_path_buf(),
+                target_version,
+                ledger_db_batch_size: 2,
+            };
+
+            cmd.run().unwrap();
+
+            let (ledger_db, state_merkle_db, _) = AptosDB::open_dbs(
```

Comment:
the in-depth examination below are nice, but I'd also actually open `AptosDB` and query the last version, get txns_list_with_proof, stream a snapshot, etc, to verify the data structure is intact.

Reply:
Done